Sources/MapboxCoreNavigation/MapboxRoutingProvider.swift

```diff
@@ -0,0 +1,344 @@
+@_implementationOnly import MapboxCommon_Private
+import MapboxDirections
+import MapboxNavigationNative
+@_implementationOnly import MapboxNavigationNative_Private
+
+/**
+ Provides alternative access to routing API.
+ 
+ Use this class instead `Directions` requests wrapper to request new routes or refresh an existing one. Depending on `RouterSource`, `MapboxRoutingProvider` will use online and/or onboard routing engines. This may be used when designing purely online or offline apps, or when you need to provide best possible service regardless of internet collection.
+ */
+public class MapboxRoutingProvider: RoutingProvider {
+    
+    /**
+     Initializes new `MapboxRoutingProvider`.
+     
+     - parameter source: routing engine source to use.
+     - parameter settings: settings object, used to get credentials and cache configuration.
+     */
+    public init(_ source: Source = .hybrid, settings: NavigationSettings = .shared) {
+        self.source = source
+        self.settings = settings
+        
+        let factory = NativeHandlersFactory(tileStorePath: settings.tileStoreConfiguration.navigatorLocation.tileStoreURL?.path ?? "",
+                                            credentials: settings.directions.credentials)
+        self.router = RouterFactory.build(for: source.nativeSource,
+                                          cache: factory.cacheHandle,
+                                          historyRecorder: factory.historyRecorder)
```

Comment:
I found the following issues with this approach:
- `NativeHandlersFactory` is created without the potentially adjusted history URL. In the example app, it leads to two alive CacheHandles. 
- If the above is fixed, there will still be an issue with the Router using a different instance of the history record handler. 

I don't think we are supposed to have more than one history recorder unless it is an artificial use case. 

So we shall reuse the history recorder from `Navigator` and, in the future, decouple HistoryRecorder from Navigator/PassiveLocationProvider/RouteController.

Reply:
We assume that settings are configured before any call to SDK entities is done. Due to `CacheHandlerFactory` `CacheHandle` instance should be reused between `MapboxRoutingProvider` and `Navigator`. But history recorder will be created twice, that's true. Not sure how critical it is, and not sure what's the best way to mitigate this. Referring to `Navigator` here is not an option since it will trigger `MBNNNavigator` initialization which is not desired.

Reply:
> We assume that settings are configured before any call to SDK entities is done. Due to CacheHandlerFactory CacheHandle instance should be reused between MapboxRoutingProvider and Navigator. 

Please try Example app, it doesn't seem to work as you describe. 

> Referring to Navigator here is not an option since it will trigger MBNNNavigator initialization which is not desired.

Well, then we need similar approach to CacheHandlerFactory. Or change it completely to not only cache cachehandle, but other relevant instances as well.

Reply:
Indeed, there was a difference in cache key form Routing Provider and Navigator, I think I fixed it by adding more parameters to CacheHandlerFactory initialization. At least it should work if `Navigator`'s parameters are not modified before launch.
I also added same caching mechanism for `HistoryRecorder`. I would prefer to solve this issue in a different way, but I don't know how, and it is outside of this task's scope. Currently, only `CacheHandle` and `HistoryRecorder` instances are used in multiple places, so I've added caching only for these 2. With `HandlerFactory` it should be trivial to add caching to any other required entity though.

Reply:
Let's resolve this thread, although it is not clear what will happen during switch to offline/online for Navigator.